### DISCUSSION TITLE
UE5 editor for new Aircraft

### DISCUSSION THREAD

**Discussion by ICanDoThings2** [JSBSim-Team/jsbsim#618] [Ideas]
I think it would be great if in editor there was a way to tweak values for export of new aircraft models, rather than just constantly refreshing an .xml or so, and being able to move the points for CG and the rest. Also for where you put fuel tanks and so forth. It would just make it more intuitive in my opinion. Also examples for other aircraft and how to set up and run them would be good I think, like for helicopters and prop planes.

**Comment by seanmcleod** [comment]
@ICanDoThings2 I'm not sure if this is the best place to discuss UE5 specific ideas if it doesn't relate to the basic "glue code" that enables integration of JSBSim and UE5. For example I'm not sure whether a UE5 editor is something that 3rd parties can even develop to plug into the UE development environment or whether it's something that only Epic Games can develop/add, i.e. it's really a request to Epic Games, and if so the request should be posted on Epic Games's support forums.

Now in the past there was a program developed, JSBSim Commander - http://jsbsimcommander.sourceforge.net/ that was designed to provide the sort of GUI editing of JSBSim xml files that you envisage although it's platform agnostic.

However it looks like it's been inactive since 2011 and won't work with the current JSBSim xml files.

>  I have marked JSBSim Commanded as "Inactive". JSBSim code itself has progressed much since Commander was last updated. JSBSim Commander will not work with the latest JSBSim code and flight model configurations.*

However this repo/forum would be the place to discuss re-activating JSBSim Commander or something similar, i.e. something generally useful for all JSBSim users and not UE specific/only. 

@bcoconni , @AlbanBERGERET-Epic your thoughts?

**Reply by bcoconni** [reply]
As @seanmcleod said, improvements to the UE5 editors for JSBSim should better be hosted in some dedicated repositories. As far as I am concerned, I would gladly create another repository under the [JSBSim-Team organization](https://github.com/JSBSim-Team) umbrella that would host UE5 specific tools for JSBSim. That would be great to have a community dedicated to JSBSim and UE5 interaction in that repository :smile:.

However the current repository `JSBSim-Team/JSBSim` is focused on the development of the JSBSim library itself and its bindings with other environments: Python, Matlab, UE5, Julia (which is still WIP at the moment).

Regarding JSBSimCommander, I have copied the code to the [JSBSim-Team/jsbsimcommander](https://github.com/JSBSim-Team/jsbsimcommander) repository. Unfortunately, it uses an outdated version `2.*` of [wxWidgets](https://www.wxwidgets.org) which is incompatible with the current API version `3.*` of wxWidgets. So the development on this tool is stalled.

**Comment by ICanDoThings2** [comment]
Interesting. Perhaps re-activating it would be good. My questions and interests are more relevant to UE5 and JSBSim's integration with it. JSBSim's integration seems uncomplete, is there a roadmap of planned features and so forth? I was looking through the plugin files and a few of the engine types seemed to have To-Do under their code as comments. Also I didn't even see where it was for the rotor-wings/helicopters.

**Reply by bcoconni** [reply]
> Interesting. Perhaps re-activating it would be good.

As I mentioned above, this would need migrating JSBSimCommander from wxWidgets 2.x to wxWidgets 3.x which as far as I could investigate is not a simple task.
> My questions and interests are more relevant to UE5 and JSBSim's integration with it. JSBSim's integration seems uncomplete, is there a roadmap of planned features and so forth? I was looking through the plugin files and a few of the engine types seemed to have To-Do under their code as comments. Also I didn't even see where it was for the rotor-wings/helicopters.

No official plans at the moment, but that does not mean there could not be any. The UE5 plugin for JSBSim is an initiative from Epic Games that has just been included in JSBSim. We are glad that Epic Games has selected JSBSim as the flight dynamics library for their [Antoinette project](https://unrealengine.com/en-US/blog/antoinette-project-tools-to-create-the-next-generation-of-flight-simulators).

However, to be honest, we are quite limited in manpower so this initiative can only thrive if there is a community behind it. So we are welcoming all contributions related to the interaction between UE5 and JSBSim whether they are ideas, code, examples, documentation, etc.

**Reply by ICanDoThings2** [reply]
Well. I'm interested, but I'm not quite sure how far developed to couple with UE5 JSBSim was/is. It seemed like there was no support still for props or helicopters/rotors, I assume the code still needs further dissection and porting. What sort of unit does JSBSim use internally for distance as well, for the local frame? Meters? Centimeters? I know UE5 uses centimeter but wasn't sure about JSBSim.

**Reply by bcoconni** [reply]
> Well. I'm interested, but I'm not quite sure how far developed to couple with UE5 JSBSim was/is.

Well, the UE5 plugin has just been added so there remains a lot of room for improvement. Now is a great time to get on board since almost everything remains to be implemented.
> It seemed like there was no support still for props or helicopters/rotors, I assume the code still needs further dissection and porting.

The plugin is designed as an example so indeed lots of feature could be added, the sky is the limit :smile: 
> What sort of unit does JSBSim use internally for distance as well, for the local frame? Meters? Centimeters? I know UE5 uses centimeter but wasn't sure about JSBSim.

JSBSim uses imperial/US units as the aeronautics industry does, so internally lengths are in feet. JSBSim can read SI units from the XML files though.

**Reply by ICanDoThings2** [reply]
AH, the icky feet. How would I go about contributing? I will work on the code, but where would be the best place to share it in your opinion?

**Reply by ICanDoThings2** [reply]
I'm still waiting to hear from AlbanBERGERETEpic too, he said he was travelling last we heard from them.

**Reply by bcoconni** [reply]
> AH, the icky feet.

JSBSim has a number of C++ getter methods that output values in various units systems so this is something that can easily be worked around.
> How would I go about contributing? I will work on the code, but where would be the best place to share it in your opinion?

For a start, if you are contributing code to the plugin itself I'd suggest working with the main JSBSim repository. I don't know how familiar you are with GitHub but the standard workflow is to fork the current repository under your username then hacking in your fork and finally submitting a pull request (see [GitHub docs for contributing to projects](https://docs.github.com/en/get-started/quickstart/contributing-to-projects)).

**Reply by bcoconni** [reply]
> I'm still waiting to hear from @AlbanBERGERET-Epic too, he said he was traveling last we heard from them.

OK. No problem.

**Reply by ICanDoThings2** [reply]
> > AH, the icky feet.
> 
> JSBSim has a number of C++ getter methods that output values in various units systems so this is something that can easily be worked around.
> 
> > How would I go about contributing? I will work on the code, but where would be the best place to share it in your opinion?
> 
> For a start, if you are contributing code to the plugin itself I'd suggest working with the main JSBSim repository. I don't know how familiar you are with GitHub but the standard workflow is to fork the current repository under your username then hacking in your fork and finally submitting a pull request (see [GitHub docs for contributing to projects](https://docs.github.com/en/get-started/quickstart/contributing-to-projects)).

Alright. I am vaguely familiar with Git/GitHub, but wasn't sure if I needed permissions or something to do a fork. Pull requests make sense then since it is approved before being done.

**Reply by bcoconni** [reply]
You just need a GitHub account to fork any GitHub public project such as JSBSim. Once forked, you own the forked repository and you are allowed to do whatever you deem relevant with it. As you mentioned, pull requests go through a review process before including your contributions to the official repository.

**Reply by ICanDoThings2** [reply]
Right. But I think again @AlbanBERGERET-Epic might have something more coming or, maybe not but I will wait to hear in case, to avoid double work.

**Comment by seanmcleod** [comment]
> JSBSim's integration seems uncomplete, is there a roadmap of planned features and so forth?

I guess it depends on what your expectations are versus what Epic's aim is. I'm not sure that Epic, @AlbanBERGERET-Epic would need to confirm, are aiming to provide a complete turn-key system/sample.

The impression I get is that this is simply a basic reference example for people to use as a starting point if they are interested or thinking of implementing a flight simulator using UE.

https://dev.epicgames.com/community/learning/tutorials/mmL/a-diy-flight-simulator-tutorial

> With the publication of [Antoinette Project](https://unrealengine.com/en-US/blog/antoinette-project-tools-to-create-the-next-generation-of-flight-simulators), we wanted to illustrate how simple and fast it is to assemble various buildings blocks and create a rough sketch of a flight simulator.

So a "rough sketch", so my impression is they're not planning a fully detailed, fully comprehensive sample using JSBSim, but I could be wrong.

**Comment by ICanDoThings2** [comment]
Ah. I wasn't sure, It didn't seem clear. But if that's the case, it's good to know. I hope it won't be too hard to work further off the example to add functionality with UE5. It seems like most of the work has been done though.

**Comment by AlbanBERGERET-Epic** [comment]
Hey guys,

I don't have time yet because I'm traveling, but I saw your posts and I'll
provide you with a more elaborate answer (and some extra ideas) later...

Keep posted!

**Reply by ICanDoThings2** [reply]
Wonderful! I hope the wonderful things we've got so far from you and UE5 integration with JSBSim won't be the last! I was actually working on a port of it myself, and it was fantastic to have someone at least done some of it so I am hoping there will be more to work with.

**Comment by ICanDoThings2** [comment]
This is also a question about the base JSBSim though it is related; can the weight of an aircraft be changed, increased or decreased at runtime? Like I saw there was a PointMass which sounds like it might do something like that.

**Comment by AlbanBERGERET-Epic** [comment]
Hi everyone, 

I'm finally back in the office, so I can give you my opinion on all of your questions above. 

--- A little bit of context ---

We launched Project Antoinette to demonstrate to the Flight Simulation Industry that contrary to former generation Games Engines, UE5 was ready to empower flight simulators (Double Precision, Large open worlds, Volumetric Clouds, Georeferencing, Performance, Realism, etc). Moreover, historically, the simulation industry made the distinction between the Simulation (runs the models) and the Image Generators (Produces pixels). But today, a modern game engine can do both! And even if the default internal physic engine is aimed at entertainment for gaming, the co-simulation concept can be used to host any specific and serious simulation model directly inside the Engine. C++ has always been the language of choice for simulation, so integrating these models or libraries is straightforward. 

And because we are more doers than speakers, we did a demo for that...

But what to do with the outcome? In accordance with our open-source philosophy, we decided to share it with the community. To prove that it was not smoke and mirrors, but also to share knowledge and stimulate this community of fans. Thanks to the JSBSim team for welcoming this work into your repository! 

Now, what about the future. I totally understand your legitimate question such as "Will Epic Games release a Flight Simulator in 6 months?". 
I can't comment on Epic Games' roadmap, but what I can tell is that this sample is an initiative of our Simulation team, totally unrelated to what we do in our gaming division. My personal standpoint is that we can't do every game in the world and that we are already so incredibly busy with Fortnite and Unreal Engine development that I doubt we'll have the resources for it. (And the target audience is not on the same scale).

What I can say though, is that it's not in our goal to provide turn-key systems. We want to provide tools for creators, to make their life easier in developing games, simulations, or other 3D interactive content. We hope you'll enjoy doing it and even sometimes getting money from it. We don't want to do everything and eat your food! If a common feature is requested by a large community, it will probably end up in the engine platform. But for all specific applications, that is up to you folks to work... (Otherwise we would have to hire a crazy amount of people to fulfill all the needs!)

--- UE5 as an editor ---

Well, it depends if you are searching for a UI or a workflow... 

If searching for a UI, I would recommend a basic external editor made in Qt or Winform/WPF with XML Serialization. But we must be pragmatic. Is it worth it? It takes a lot of time to create and maintain such a tool... Compared to the number of aircraft to be authored, it has to be balanced with the time it takes with a simple text editor to directly edit the XML... 

If you are searching for a workflow, more iterative, the way to proceed in UE would be to work with component properties or better, with assets. You can create your custom assets in UE (like textures, materials, meshes), and have an editor UE associated with it. [This blog post](https://gmpreussner.com/reference/adding-new-asset-types-to-ue4) is well done to know the basics.

One could imagine having a library of assets made in a such way. But that will come with some drawbacks too: A dependency on Unreal, assets are binary, so it would require locking mechanisms for source control. But it would also have some advantages, like having access to UE Editors (basic types, arrays, structs, curves), being able to test on the fly in the app, and the ability to import/export to XML if needed. 
I'm also not sure about how the generic data repository model of JSBSim and FlightGear would go through that. 

I agree with @bcoconni that the JSBSim-Team/JSBSim repo is for JSBSim code and its XML Data. UEReferenceApp is a kind of test app to check for regressions in the UE plugin code. We should consider other targets for JSBSim like Python, and Mathlab. (even if I consider a 3D view helps a lot sometimes!). If one would like to do a better flight sim application or additional extended model library (with models, sounds, and systems like we can find in FlightGear), it would be better somewhere else. 

--- Roadmap ---

Ouch, you saw them! Yeah, I know, there are some TODOs in my code, like missing engine types. 
It was because I mainly focused on the demo using an airliner, but I'm pretty sure it's not that long to add. 
Having a full coverage of engine types, and added additional commands (I skipped spoilers and reverse thrust because of the short deadline) should be easy. 
As you can imagine, we have plenty of other stuff running at Epic, and my days are really (really) busy. Even if I can help writing some additions from time to time, it will be hard for me to commit to that. 
But if one day someone is willing to gather folks and create a "Next-Gen Open Source Flight sim project" with a roadmap and a crew, I'll be happy to contribute. Maybe not by coding, but at least by providing guidance and advice on the UE5 side of things. (eg: Don't work on that feature, it's coming in our next version...) 

It really depends on my work targets. If one day I have to demonstrate something to a helicopter company, I would for sure add this to the plugin. And remember the application was simple. I wrote it was a "rough sketch". No cockpit, no internal systems, no VR... All of that is possible with UE, but a solid and coordinated crew would be needed to bring that to life.

I'll be always happy to chat about it and help you as much as I can. So don't hesitate to quote me with any questions you could have

Alban

**Reply by ICanDoThings2** [reply]
Alright. I was just checking you weren't working on more. Didn't want to accidently do double work. But, if it's done for now, I guess I will poke around more in it. So to be clear; no updates coming for the UE5 plugin?

**Reply by AlbanBERGERET-Epic** [reply]
No, nothing is planned for the immediate future. I have some travels ahead, so I don't plan to come back on it soon. You're good to go. (And I'll get your additions with great pleasure). 
Just a quick notice: Be careful not to publish in the repo any personal data. Eg: If for instance you use Cesium to get a better terrain as explained in the DIY tutorial, make sure not to expose your token!

**Reply by ICanDoThings2** [reply]
I will be careful. I am also reconsidering making my own plugin, more tightly coupled to UE5 rather than just a basic import for the most part like it seems you did, no offense. But, I think you still provide insight with the plugin you gave as well. I will look through it more in either case.

**Reply by AlbanBERGERET-Epic** [reply]
What do you mean by "more tightly coupled with UE5"? The idea of using it as an editor for model properties? 
I did not want to go this way and keep the JSBSim soft+data encapsulation. But this is a matter of design choice...

**Reply by ICanDoThings2** [reply]
I had some design ideas I wanted. They're just mine, mind you. It would still be a plugin even for what I would do. I was just trying to trim the fat, cut variables that weren't needed and could go off what could be in UE5 and all. I will look through your plugin more though, and if it needs less tweaking to get where I need it to be I will probably go off it instead. I just don't have enough facts yet. I also don't know if JSBSim has functions I need it to either, like disabling engines (I can just cut throttle though and prevent it from being raised), as well as adding/removing/changing weight from things added or removed to the aircraft, for instance.
